spec/statement.dd

```diff
@@ -698,88 +698,100 @@ foreach (string s, double d; aa)
 
     $(SPEC_RUNNABLE_EXAMPLE_COMPILE
     ---
     struct S
     {
-        int opApply(scope int delegate(ref uint n) dg);
+        uint n;
+        int opApply(scope int delegate(ref uint) body) => body(n);
     }
-    void f(S s)
+    void main()
     {
+		S s;
         foreach (ref uint i; s)
-            i++;
+        {
+            i++; // effectively s.n++
+        }
+        foreach (i; s)
+        {
+            static assert(is(typeof(i) == uint));
+            assert(i == 1);
+        }
     }
     ---
     )
-    Above, `opApply` is still matched when `i` is not `ref`, so by using
-    a `ref` delegate parameter both forms are supported.
-    )
+    Above, `opApply` is matched both when `i` is and is not `ref`, so by using
+    a `ref` delegate parameter, both forms are supported.
 
-    $(P There can be multiple $(D opApply) and $(D opApplyReverse) functions -
-        one is selected
-        by matching each parameter of `dg` to each $(I ForeachType)
-        declared in the $(I ForeachStatement).)
+    Stating the type of the variable like in the first loop is optional,
+    as is showcased in the second loop, where the type is inferred.
+    )
 
-    $(P The body of the apply
-        function iterates over the elements it aggregates, passing each one
-        in successive calls to the `dg` delegate. The delegate return value
-        determines whether to interrupt iteration:)
+    $(P The apply
+        function iterates over the elements it aggregates by passing each one
+        in successive calls to the `body` delegate. The delegate return value
+        of the `body` delegate
+        determines how to proceed iteration:)
 
     $(UL
-        $(LI If the result is nonzero, apply must cease
+        $(LI If the result is nonzero, the apply function must cease
         iterating and return that value.)
-        $(LI If the result is 0, then iteration should continue.
+        $(LI If the result is `0`, then iteration should continue.
         If there are no more elements to iterate,
-        apply must return 0.)
+        the apply function must return `0`.)
     )
 
-    $(P The result of calling the delegate will be nonzero if the *ForeachStatement*
+    $(P The result of the `body` delegate will be nonzero if the *ForeachStatement*
         body executes a matching $(GLINK BreakStatement), $(GLINK ReturnStatement), or
         $(GLINK GotoStatement) whose matching label is outside the *ForeachStatement*.
     )
 
-    $(P For example, consider a class that is a container for two elements:)
+    $(P The $(D opApply) and $(D opApplyReverse) member functions can be overloaded.
+        Selection works similar to overload resolution by
+        comparing the number of `foreach` variables and number of parameters of `body` and,
+        if more than one overload remains unique overload,
+        matching the parameter types of `body` and each $(I ForeachType)
+        declared in the $(I ForeachStatement) when given.)
```

Comment:
```suggestion
        comparing the number of `foreach` variables with the number of parameters of `body` and,
        if more than one overload remains,
        matching each parameter of `body` against each $(I ForeachType)
        declared in the $(I ForeachStatement).)
```
Note: I've used *each parameter* rather than *each parameter type* because `ref` affects overload resolution as well as the type.